File: sdk-go/pace/logger.go
package pace

import (
	"fmt"
	"os"
	"strings"
)

func normalizeLogMode(mode string, debug bool) LogMode {
	switch strings.ToLower(mode) {
	case "pretty":
		return LogPretty
	case "compact":
		return LogCompact
	case "off":
		return LogOff
	default:
		if debug {
			return LogCompact
		}
		return LogOff
	}
}

func shouldIgnoreRoute(route string) bool {
	return route == "/favicon.ico" ||
		route == "/health" ||
		route == "/api/health" ||
		route == "/healthz" ||
		route == "/ping" ||
		route == "/api/ping" ||
		strings.HasPrefix(route, "/_next")
}

func formatDurationMs(ms int64) string {
	if ms <= 0 {
		return "0ms"
	}
	if ms%3_600_000 == 0 {
		return fmt.Sprintf("%dh", ms/3_600_000)
	}
	if ms%60_000 == 0 {
		return fmt.Sprintf("%dm", ms/60_000)
	}
	if ms%1000 == 0 {
		return fmt.Sprintf("%ds", ms/1000)
	}
	return fmt.Sprintf("%dms", ms)
}

func logDecision(mode LogMode, decision CanonicalDecision) {
	if mode == LogOff || shouldIgnoreRoute(decision.Route) {
		return
	}

	headline := "BLOCK"
	switch decision.Decision {
	case DecisionAllow:
		headline = "ALLOW"
	case DecisionWouldBlock:
		headline = "WOULD_BLOCK"
	}

	if mode == LogCompact {
		parts := []string{fmt.Sprintf("[Pace] %s", headline)}
		switch decision.Decision {
		case DecisionAllow:
			parts = append(parts, fmt.Sprintf("route=%s", decision.Route))
			if decision.Remaining >= 0 {
				parts = append(parts, fmt.Sprintf("remaining=%d", decision.Remaining))
			}
			if decision.LatencyMs >= 0 {
				parts = append(parts, fmt.Sprintf("latency=%dms", decision.LatencyMs))
			}
		case DecisionWouldBlock:
			parts = append(parts, decision.Route, fmt.Sprintf("reason=%s", decision.Reason))
			if decision.RefillMs > 0 {
				parts = append(parts, fmt.Sprintf("refill=%s", formatDurationMs(decision.RefillMs)))
			}
		default:
			parts = append(parts, decision.Route, fmt.Sprintf("reason=%s", decision.Reason))
			if decision.ResetMs > 0 {
				parts = append(parts, fmt.Sprintf("reset=%s", formatDurationMs(decision.ResetMs)))
			}
			if decision.RefillMs > 0 {
				parts = append(parts, fmt.Sprintf("refill=%s", formatDurationMs(decision.RefillMs)))
			}
			if decision.LatencyMs >= 0 {
				parts = append(parts, fmt.Sprintf("latency=%dms", decision.LatencyMs))
			}
		}
		_, _ = os.Stdout.WriteString(strings.Join(parts, " ") + "\n")
		return
	}

	lines := []string{fmt.Sprintf("[Pace] %s", headline)}
	if decision.Decision == DecisionAllow {
		lines = append(lines, "", fmt.Sprintf("Route: %s", decision.Route))
		if decision.Remaining >= 0 {
			lines = append(lines, fmt.Sprintf("Remaining: %d", decision.Remaining))
		}
		if decision.LatencyMs >= 0 {
			lines = append(lines, fmt.Sprintf("Latency: %dms", decision.LatencyMs))
		}
	} else if decision.Decision == DecisionWouldBlock {
		lines = append(lines, "", fmt.Sprintf("Route: %s", decision.Route), fmt.Sprintf("Reason: %s", decision.Reason), fmt.Sprintf("Mode: %s", decision.Mode))
		if decision.RefillMs > 0 {
			lines = append(lines, fmt.Sprintf("Refill: %s", formatDurationMs(decision.RefillMs)))
		}
	} else {
		ip := decision.IP
		if ip == "" {
			ip = "unknown"
		}
		lines = append(lines, "", fmt.Sprintf("Route: %s", decision.Route), fmt.Sprintf("IP: %s", ip), fmt.Sprintf("Algorithm: %s", decision.Algorithm), fmt.Sprintf("Reason: %s", decision.Reason), fmt.Sprintf("Mode: %s", decision.Mode))
		if decision.Remaining >= 0 {
			lines = append(lines, fmt.Sprintf("Remaining: %d", decision.Remaining))
		}
		if decision.ResetMs > 0 {
			lines = append(lines, fmt.Sprintf("Reset: %s", formatDurationMs(decision.ResetMs)))
		}
		if decision.RefillMs > 0 {
			lines = append(lines, fmt.Sprintf("Refill: %s", formatDurationMs(decision.RefillMs)))
		}
		if decision.LatencyMs >= 0 {
			lines = append(lines, fmt.Sprintf("Latency: %dms", decision.LatencyMs))
		}
	}

	_, _ = os.Stdout.WriteString(strings.Join(lines, "\n") + "\n")
}

File: sdk-rust/src/logger.rs
use pace_engine::core::canonical::{CanonicalDecision, TrafficDecision};

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum LogMode {
    Off,
    Compact,
    Pretty,
}

pub fn normalize_log_mode(log_mode: Option<&str>, debug: bool) -> LogMode {
    match log_mode {
        Some("pretty") => LogMode::Pretty,
        Some("compact") => LogMode::Compact,
        Some("off") => LogMode::Off,
        Some(_) => LogMode::Compact,
        None if debug => LogMode::Compact,
        None => LogMode::Off,
    }
}

pub fn should_ignore_route(route: &str) -> bool {
    matches!(
        route,
        "/favicon.ico" | "/health" | "/api/health" | "/healthz" | "/ping" | "/api/ping"
    ) || route.starts_with("/_next")
}

fn format_duration_ms(ms: u64) -> String {
    if ms == 0 {
        return "0ms".to_string();
    }
    if ms % 3_600_000 == 0 {
        return format!("{}h", ms / 3_600_000);
    }
    if ms % 60_000 == 0 {
        return format!("{}m", ms / 60_000);
    }
    if ms % 1000 == 0 {
        return format!("{}s", ms / 1000);
    }
    format!("{}ms", ms)
}

pub fn log_decision(mode: LogMode, decision: &CanonicalDecision) {
    if mode == LogMode::Off || should_ignore_route(&decision.route) {
        return;
    }

    let headline = match decision.decision.as_str() {
        "allow" => "ALLOW",
        "would_block" => "WOULD_BLOCK",
        _ => "BLOCK",
    };

    if mode == LogMode::Compact {
        let mut parts = vec![format!("[Pace] {}", headline)];
        if decision.decision == TrafficDecision::Allow.as_str() {
            parts.push(format!("route={}", decision.route));
            if let Some(remaining) = decision.remaining {
                parts.push(format!("remaining={}", remaining));
            }
            if let Some(latency_ms) = decision.latency_ms {
                parts.push(format!("latency={}ms", latency_ms));
            }
        } else if decision.decision == TrafficDecision::WouldBlock.as_str() {
            parts.push(decision.route.clone());
            parts.push(format!("reason={}", decision.reason));
            if let Some(refill_ms) = decision.refill_ms {
                parts.push(format!("refill={}", format_duration_ms(refill_ms)));
            }
        } else {
            parts.push(decision.route.clone());
            parts.push(format!("reason={}", decision.reason));
            if let Some(reset_ms) = decision.reset_ms {
                parts.push(format!("reset={}", format_duration_ms(reset_ms)));
            }
            if let Some(refill_ms) = decision.refill_ms {
                parts.push(format!("refill={}", format_duration_ms(refill_ms)));
            }
            if let Some(latency_ms) = decision.latency_ms {
                parts.push(format!("latency={}ms", latency_ms));
            }
        }
        println!("{}", parts.join(" "));
        return;
    }

    let mut lines = vec![format!("[Pace] {}", headline)];
    if decision.decision == TrafficDecision::Allow.as_str() {
        lines.push(String::new());
        lines.push(format!("Route: {}", decision.route));
        if let Some(remaining) = decision.remaining {
            lines.push(format!("Remaining: {}", remaining));
        }
        if let Some(latency_ms) = decision.latency_ms {
            lines.push(format!("Latency: {}ms", latency_ms));
        }
    } else if decision.decision == TrafficDecision::WouldBlock.as_str() {
        lines.push(String::new());
        lines.push(format!("Route: {}", decision.route));
        lines.push(format!("Reason: {}", decision.reason));
        lines.push(format!("Mode: {}", decision.mode));
        if let Some(refill_ms) = decision.refill_ms {
            lines.push(format!("Refill: {}", format_duration_ms(refill_ms)));
        }
    } else {
        lines.push(String::new());
        lines.push(format!("Route: {}", decision.route));
        lines.push(format!("IP: {}", decision.ip.as_deref().unwrap_or("unknown")));
        lines.push(format!("Algorithm: {}", decision.algorithm));
        lines.push(format!("Reason: {}", decision.reason));
        lines.push(format!("Mode: {}", decision.mode));
        if let Some(remaining) = decision.remaining {
            lines.push(format!("Remaining: {}", remaining));
        }
        if let Some(reset_ms) = decision.reset_ms {
            lines.push(format!("Reset: {}", format_duration_ms(reset_ms)));
        }
        if let Some(refill_ms) = decision.refill_ms {
            lines.push(format!("Refill: {}", format_duration_ms(refill_ms)));
        }
        if let Some(latency_ms) = decision.latency_ms {
            lines.push(format!("Latency: {}ms", latency_ms));
        }
    }

    println!("{}", lines.join("\n"));
}